proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,385 @@
+    nonmutating set
+  }
+    
+  var projectedValue: Self {
+    self
+  }
+}
+
+typealias A = (Reference<Int>) -> Void
+
+let a: A = { @Reference foo in
+  ...
+}
+```
+
+    Becomes:
+
+```swift
+let a: A = { _foo in
+  var foo: Int {
+    get { 
+      _foo.wrappedValue 
+    }
+    set { 
+      _foo.wrappedValue = newValue
+    }
+  }
+
+  var $foo: Int {
+    get { 
+      _foo.projectedValue 
+    }
+  }
+    
+  ...
+}
+```
+
+2. Value Semantics Wrapper
+
+```swift
+@propertyWrapper
+struct Wrapper<Value> {
+  init(wrappedValue: Value) 
+    
+  var wrappedValue: Value {
+    get 
+    set
+  }
+}
+
+typealias B = (inout Wrapper<Int>) -> Void
+
+let b: B = { @Wrapper foo in
+  ...
+}
+```
+
+    Becomes:
+
+```swift
+let b: B = { _foo in
+  var foo: Int {
+    get { 
+      _foo.wrappedValue 
+    }
+    set { 
+      _foo.wrappedValue = newValue
+    }
+    // Since the paramter is marked `inout`
+    // we are allowed to have a mutating setter.
+  }
+  
+  ...
+}
+```
+
+3. Value Semantics Wrapper with Special Initializer
+
+```swift
+@propertyWrapper
+class WrapperObject<Value> {
+  init(wrappedValue: Value) 
+    
+  var wrappedValue: Value {
+    get 
+    set 
+    // Not actually mutating
+    // because `WrapperObject` is
+    // a class.
+  }
+}
+
+typealias C = (WrapperObject<Int>) -> Void
+
+let c: C = { @WrapperObject foo in
+  ...
+}
+```
+
+    Becomes:
+
+```swift
+let c: C = { _foo in
+  var foo: Int {
+    get { 
+      _foo.wrappedValue 
+    }
+    set { 
+      _foo.wrappedValue = newValue
+    }
+    // Since `WrapperObject` has reference
+    // semantics we can include a setter.
+  }
+
+  ...
+}
+```
+
+
+## Source compatibility
+
+This is an additive change with _no_ impact on **source compatibility**.
+
+
+## Effect on ABI stability
+
+This is an additive change with _no_ impact on **ABI stability**.
+
+
+## Effect on API resilience
+
+This is an additive change with _no_ impact on **API resilience**.
+
+
+## Alternatives Considered
+
+### Infer Property Wrappers in Closure Parameters 
+
+TBD
+
+
+### Support `@autoclosure` and `@escaping` in Function Parameters
+
+TBD
+
+
+## Future Directions
+
+### Support property wrapper initialization from a projected value
+
+Today, a property wrapper can be initialized from an instance of its `wrappedValue` type if the wrapper provides a suitable `init(wrappedValue:)`. The same initialization strategy is used in this proposal for property wrapper parameters to allow users to pass a wrapped value as a property wrapper argument. We could extend this model to support initializing a property wrapper from an instance of its `projectedValue` type by allowing property wrappers to define an `init(projectedValue:)` that follows the same rules as `init(wrappedValue:)`. This could allow users to additionally pass a projected value as a property wrapper argument, like so:
+
+```swift
+@propertyWrapper
+struct Wrapper<T> {
+  var wrappedValue: T
+  var projectedValue: ProjectionType
```

Comment:
Should we maybe specify the projection type, like this:

```swift
@propertyWrapper
struct Wrapper<Value> {

  typealias ProjectionType = ...

  var projectedValue: ProjectionType 


  init(projectedValue: ProjectionType) { ... }

  
  var wrappedValue: Value

}
```

I think we should include the type alias so that the reader doesn't have to search in the proposal to find if we refer to something specified previously. Instead, they'll see the type alias and understand that the actual type is irrelevant without being uncertain of whether it's been specified. 

> **NOTE**: The other changes to layout are just personal choice; feel free to leave them out.

Reply:
You're totally right that this example isn't clear. Do you think using the same example throughout the proposal and building on it would clear up the confusion? We could still include a repeated definition of whichever wrapper type we choose so that the reader doesn't have to jump back and forth between different parts of the proposal. Also, using a more realistic example would better illustrate what these future directions enable in terms of expressivity instead of just showing the mechanics of how it would work

Reply:
Using real-world wrappers (but maybe with simplified implementations) seems like a really good idea and I like how exploring future directions through such wrappers emphasizes their usefulness. As you can imagine, though, we won't be able to use just one wrapper in all the sections. However, I think we could stick to `@Clamped` for the Motivation and whatever other section qualifies. One such section, for example, is Detailed Design where we could use `@Clamped` instead of `@Wrapper` is some examples and of course it some Future Direction sub-sections would also apply - as you mentioned.

Reply:
I've change the example in the Future Direction section you wrote to this:

---

```swift
@propertyWrapper
struct Clamped<Value: Comparable> {
  ...

  init(projectedValue: Self) { ... }
}

func distanceFromUpperBound(
  @Clamped clamped: Int
) { ... }

distanceFromUpperBound(
  $clamped: Clamped(to: 0 ... 100, wrappedValue: 30)
) // returns: 70
```

---

It's something that one could actually see in the real world. Anyway, let me know if you like this better or if you prefer that we revert to the previous one.